### PR TITLE
[controller] Add InteractionPeer interface for cluster control

### PR DESCRIPTION
**Pull request by gjc13**
#### Problem

The `controller::Device` class is currently tightly bond with `controller::DeviceController` and cannot be used for device-to-device interaction.

#### Change overview

* Split all cluster interaction functions to `InteractionPeer` to decouple them from commission and storage.

#### Testing

Manually build and run. The chip-tool pairing and control is working.

### CODE REVIEW

**Review by woody-apple — COMMENTED** [review]
I'm not quite clear what the change here is, and what use case doesn't work? Can you please help clarify? The diffs don't seem to support what the description of the change is...

**Comment by gjc13** [comment]
> I'm not quite clear what the change here is, and what use case doesn't work? Can you please help clarify? The diffs don't seem to support what the description of the change is...

The use case is we'd like to utilize the `Device` class on an end device to send ZCL messages to other devices.
Thus the exchange manager, key id allocator etc. shall come from `app/server` so we added the `GetGlobal` functions.
The `MarkOperationalCertProvisioned` function is added since currently the `Device` class requires the `DeviceController` to drive its state. For controlling other devices we don't need to setup the PASE connection so this function is added to drive the state manually.

File: src/controller/CHIPDevice.h
Comment (tcarmelveilleux) [inline src/controller/CHIPDevice.h:376]: Why is this needed? It's not called

File: src/app/server/Server.cpp
Comment (tcarmelveilleux) [inline src/app/server/Server.cpp:519]: Is there a cleaner way than just having accessors for singletons? Seems like we could inject there rather than have more singleton accessors.

File: src/controller/CHIPDevice.h
Comment (gjc13) [reply src/controller/CHIPDevice.h:376]: It's used for working around assumed comissioning. The controller example is currently in inner projects.

The usage will be like:
```
        SecureSessionMgr &sessionMgr = GetGlobalSessionManager();
        chip::Messaging::ExchangeManager &exchangeMgr = GetGlobalExchangeManager();
        chip::Controller::ControllerDeviceInitParams params = {nullptr,          &sessionMgr,
                                                               &exchangeMgr,     nullptr,
                                                               nullptr,          &mOperationalCredentialSet,
                                                               mCredentialIndex, &GetGlobalSessionIDAllocator(),
                                                               nullptr};
        mDevice.Init(params, CHIP_PORT, mPeerNodeId, mPeerAddr, mFabricIndex);
        mDevice.MarkOperationalCertProvisioned();
```
As to refactoring, my idea is to make CASE peer a seperate class and decouple it with commissiong and persistence. Does it look fine to you?

File: src/app/server/Server.cpp
Comment (gjc13) [reply src/app/server/Server.cpp:519]: Can you give a detailed description on the injection? Thanks.

File: src/controller/CHIPDevice.h
Comment (gjc13) [reply src/controller/CHIPDevice.h:376]: Adding @pan-apple as the origin author of the code.

**Comment by gjc13** [comment]
@tcarmelveilleux Code refactored. The cluster interaction functionality is moved to a separated class so that device-to-device interaction an extend from it.

File: src/app/server/Server.cpp
Comment (tcarmelveilleux) [reply src/app/server/Server.cpp:519]: In the code snippet you have:

```
   SecureSessionMgr &sessionMgr = GetGlobalSessionManager();
        chip::Messaging::ExchangeManager &exchangeMgr = GetGlobalExchangeManager();
        chip::Controller::ControllerDeviceInitParams params = {nullptr,          &sessionMgr,
                                                               &exchangeMgr,     nullptr,
                                                               nullptr,          &mOperationalCredentialSet,
                                                               mCredentialIndex, &GetGlobalSessionIDAllocator(),
                                                               nullptr};
        mDevice.Init(params, CHIP_PORT, mPeerNodeId, mPeerAddr, mFabricIndex);
        mDevice.MarkOperationalCertProvisioned();
```

Curious why:

1. GetGlobal* must be called, rather than already having access to the object references at the call site (e.g. previously injected dependency)
2. `mDevice.MarkOperationalCertProvisioned()` is a workaround around the fact that CHIPDeviceController has too tight a coupling to the commissioning cases. Suggest refactoring the underlying issue, rather than "pretending a commisisoning was done" when the class already has too many ties to some disjoint flows

File: src/controller/CHIPDevice.h
Comment (tcarmelveilleux) [inline src/controller/CHIPDevice.h:351]: Typo on `Session` in `GetSessssionHandle`.

This assumes a single session per device (which may be correct, but we are not addressing the lifecycle of that session in the change here)

File: src/controller/InteractionPeer.cpp
Comment (tcarmelveilleux) [inline src/controller/InteractionPeer.cpp:42]: nit: Typo in `applicaiton`

File: src/controller/InteractionPeer.cpp
Comment (tcarmelveilleux) [inline src/controller/InteractionPeer.cpp:45]: Could the InteractionModelEngine instance have been passed in the constructor of the  InteractionPeer ?

File: src/controller/InteractionPeer.cpp
Comment (tcarmelveilleux) [inline src/controller/InteractionPeer.cpp:61]: Could this block? What is done in PreparePeer innards?

File: src/controller/InteractionPeer.cpp
Comment (tcarmelveilleux) [inline src/controller/InteractionPeer.cpp:65]: If AddResponseHandler supports nullptr in either onSuccess/onFailureCallback, the outer `if` is not needed.

File: src/controller/InteractionPeer.cpp
Comment (tcarmelveilleux) [inline src/controller/InteractionPeer.cpp:91]: @pan-apple this is a multi-admin issue I think, right here. The fact that NodeID uniqueness is globally expected across different fabrics for callback tracking properly.

File: src/controller/InteractionPeer.h
Comment (tcarmelveilleux) [inline src/controller/InteractionPeer.h:67]: Need to explain what this class is meant to do, and what the different methods do.

File: src/controller/InteractionPeer.cpp
Comment (gjc13) [reply src/controller/InteractionPeer.cpp:45]: Removed and fetch the object when used.

File: src/controller/InteractionPeer.cpp
Comment (gjc13) [reply src/controller/InteractionPeer.cpp:61]: `PreparePeer` won't block. The `Device` class logic remains the same after refactoring. It calls `NewPairing` to load the key for the peer.

File: src/controller/InteractionPeer.cpp
Comment (gjc13) [reply src/controller/InteractionPeer.cpp:91]: This function is moved from the `CHIPDevice` class. I believe it's okay to have the comment here.

File: src/app/server/Server.cpp
Comment (gjc13) [reply src/app/server/Server.cpp:519]: The `GetGlobal*` class are added since these objects are located in `app/server` and it's hard to determine the device to be added there. For instance we may want to scan for a device first and then create a control session.

File: src/controller/CHIPDevice.h
Comment (gjc13) [reply src/controller/CHIPDevice.h:351]: As to the lifecycle of a session, it's supposed to be a simple combination of {fabric-id, node-id, key-id} and be passed by value or const reference. Unfortunately the current interaction model code are accepting pointers and use `nullptr` to create default values. It's tracked in https://github.com/project-chip/connectedhomeip/issues/6792. Fixing this will require another refactor in the `app/` directory and is beyond the scope of this PR.

File: src/controller/InteractionPeer.h
Comment (gjc13) [reply src/controller/InteractionPeer.h]: Doc string added. These functions are moved from the `Device` class. I'm not very sure about the functionality of some methods. TODOs added for the origional author @erjiaqing to add doc strings.

File: src/app/server/Server.cpp
Comment (holbrookt) [reply src/app/server/Server.cpp:534]: I think the issue here is that Server.cpp instantiates its own singleton classes like `ExchangeManager`. Instead, we could initialize Server.cpp with an `ExchangeManager` (and other singletons) and those would then be accessible for the application to use to initialize a `Device`. Is this aligned with what you're suggesting @tcarmelveilleux ?

File: src/controller/CHIPDevice.cpp
Comment (holbrookt) [inline src/controller/CHIPDevice.cpp:635]: I see that `PreparePeer()` calls `LoadSecureSessionParametersIfNeeded()`: does this require that a secure session was already established at some point by the device that is the target of the connection here?

For example, would the following be valid with your change?
1. device A is commissioned by device B (device A establishes SecureSession to B)
2. commissioning SecureSession is closed
3. device B establishes new SecureSession to A using your changes

Comment (gjc13) [reply src/controller/CHIPDevice.cpp:635]: Yes, a new `InteractionPeerDelegate` implementation can be added where a CASE session is setup with the provisioned operational credential.

**Comment by gjc13** [comment]
@tcarmelveilleux @woody-apple Friendly reminder

File: src/app/server/Server.cpp
Comment (gjc13) [reply src/app/server/Server.cpp:534]: Moving these singletons are possible but requires broad changes in the examples. I'll remove the `GetGlobal` changes from this PR and add a separate PR to resolve this issue.

**Review by woody-apple — DISMISSED** [review]
Can we add tests that failed before, but pass now?

File: src/app/server/Server.cpp
Comment (andy31415) [reply src/app/server/Server.cpp]: I believe dependency injection should be vastly preferred to global singletons. Singletons cause tight coupling and make testing a lot more difficult.

File: src/controller/InteractionPeer.h
Comment (msandstedt) [inline src/controller/InteractionPeer.h:29]: The connection to a peer is a session.  Why can't we just use `CASESession` and `PASESession` objects to provide this abstraction?  These already implement the `Messaging::ExchangeDelegate` interface.  Isn't that fundamentally what's needed?

File: src/app/server/Server.cpp
Comment (gjc13) [reply src/app/server/Server.cpp]: Moved to https://github.com/project-chip/connectedhomeip/pull/9200

File: src/controller/InteractionPeer.h
Comment (gjc13) [reply src/controller/InteractionPeer.h:29]: Using the CASESession isn't enough since the `CASESession` class is more about authenticating a device with its certification. After that we need to call `SecureSessionMgr::NewPairing` which in turn calls `CASESession::DeriveSecureSession`. When talking to the device afterwards, we are not using the `CASESession` object but creating a new `ExchangeContext` instead.

**Comment by gjc13** [comment]
> Can we add tests that failed before, but pass now?

The PR is built for adding data model clients on devices and needs to work with https://github.com/project-chip/connectedhomeip/pull/9200. After the two PRs are merged the switch device can be an example usage. We can add tests with the switch.

File: src/controller/InteractionPeer.h
Comment (msandstedt) [reply src/controller/InteractionPeer.h:29]: It's not the fact these objects require additional lifecycle prodding that precludes their use in the way I describe.  It's the fact that the stack does not directly use CASESession and PASESession objects to send messages.

Instead, the secure session manager provides a SessionHandle that can be used to construct exchange contexts, which then allows a key for matching messages traveling back and forth between the exchange manager and session manager singletons.  It's not how I would do it, but I see now this is the code does it.

Why then is SessionHandle not more fundamental to the design?  Why is this only accessible through the `InteractionPeerDelegate`?  I would think we would be storing this on construction of the `InteractionPeer`, or even extending from it, since it's the fundamental identifier of the InteractionPeer.  Why do we need to register a delegate to hold state for us?  Can it ever change?  If it did, would that be correct behavior?

And if we don't have an underlying Device object, what do we use as a delegate then?

Comment (gjc13) [reply src/controller/InteractionPeer.h:29]: I believe you are correct and the centralized object is the `ExchangeContext`. `SessionHandle` is just a tuple used to create it. The challenge lies in two points:

* Our stack doesn't use `ExchangeContext` as the only handle to send and receive messages. The generated cluster code uses `deviceID`. The interaction model is using `Sessionhandle`. Moreover, we have `SendCommand` and `SendCommands` in `data_model/gen` and to my surprise they leads to different ExchangeContexts. This leads to the extra functions in ` InteractionPeerDelegate` like `GetDeviceID`, `GetSessionHandle` etc.
* The current `chip::Controller::Device` provides more functionality than a `ExchangeContext`. It will implicitly deserialize the object, setup the `CASESession` and even reset the whole transport layer. This leads to the `PreparePeer` function which does these work.

The `InteractionPeerDelegate` is designed so to make minimal changes to the code while decoupling the cluster client. However I believe these problems must be solved in the long run. What's the desired strategy to resolve these issues? Do we prefer to fix them once and for all in this PR or we prefer to solve it gradually?

Comment (msandstedt) [reply src/controller/InteractionPeer.h:29]: > The InteractionPeerDelegate is designed so to make minimal changes to the code while decoupling the cluster client. 

This is justification enough to me.

> However I believe these problems must be solved in the long run. What's the desired strategy to resolve these issues? Do we prefer to fix them once and for all in this PR or we prefer to solve it gradually?

I believe that if we can come to a cleaner concept of an unencrypted session and drive toward some other session-related cleanup, we may have additional opportunities for simplification and cleanup.

For now, this works and is an improvement.  I presume that if no other changes land and we want to unblock non-controller-initiated communication with this new code, we will simply need to provide something on the accessory side that can implement `InteractionPeerDelegate`.

Thanks for the background.

**Review by andy31415 — APPROVED** [review]

**Review by woody-apple — CHANGES_REQUESTED** [review]
I'm still not super clear what the added feature here is? There's a lot of new code, with no clear description of new functionality added, as well as lacking tests.

**Comment by msandstedt** [comment]
> I'm still not super clear what the added feature here is? There's a lot of new code, with no clear description of new functionality added, as well as lacking tests.

This was to provide a common interface for both controller-like devices and accessory-like devices to be able to establish CASE sessions and initiate interactions.

However, I think this may be superseded by https://github.com/project-chip/connectedhomeip/pull/9386 which provides a workaround to the Device object to allow it to be used only for CASE session establishment even if a PASE session wasn't established previously.

**Comment by gjc13** [comment]
Replaced by https://github.com/project-chip/connectedhomeip/pull/9256

**Pull request closed without merging**
